COL-378 Allow non-admin users to delete associated assets with no interactions

paulkerschen opened the pull request:
https://jira.ets.berkeley.edu/jira/browse/COL-378

nicolaasmatthijs commented on node_modules/col-assets/lib/api.js:
```diff
@@ -55,7 +55,8 @@ var UserConstants = require('col-users/lib/constants');
-    'includeWhiteboardElements': true
+    'includeWhiteboardElements': true,
+    'includeWhiteboardUsages': true
   };
```
Rather than including the whiteboard elements in which an asset is being used, can we determine whether the user can delete the asset in the back-end and just return a `canDelete` property as part of the asset profile?

paulkerschen replied:
I almost went that way, @nicolaasmatthijs, but thought that the front end might want to associate assets with whiteboards for other reasons down the line. We can move it all to the back end if you'd prefer, though.

nicolaasmatthijs replied:
That makes sense. Unfortunately, showing the active whiteboards in which an asset is used isn't all that helpful as whiteboards are private and we wouldn't be able to send the user to the whiteboard. What would be useful is to include which exported whiteboards the asset is being used in, and I'd be happy with including that in the scope of this project. In terms of the logic associated to deciding whether a user can delete an asset, I'd prefer that to be consolidated in the back-end so we can update that logic without changing the UI.

nicolaasmatthijs commented on node_modules/col-assets/lib/api.js:
```diff
@@ -144,6 +147,14 @@ var getAsset = module.exports.getAsset = function(ctx, id, opts, callback) {
       'required': false
     });
   }
+  if (opts.includeWhiteboardUsages) {
```
When an asset isn't included in an active whiteboard, it could still be part of an exported whiteboard, in which case the user should also not be able to delete it.

paulkerschen replied:
Thanks @nicolaasmatthijs. The asset_whiteboard_elements table for exported whiteboards doesn't currently have a column for the asset id of the element (as opposed to asset id of the whiteboard). It's included in the JSON but not in a way that's queryable. Probably the quickest way to make an association possible is to add a column?

nicolaasmatthijs replied:
Ah, that's unfortunate. Yeah, it's probably easiest to add a column. Note that we'll then also need to create a migration script for the existing data.

nicolaasmatthijs commented on node_modules/col-assets/lib/api.js:
```diff
@@ -824,18 +835,26 @@ var deleteAsset = module.exports.deleteAsset = function(ctx, id, callback) {
     }
 
+    // If an asset has no interactions, associated users may delete that asset. Other assets may be deleted
+    // only by course instructors
```
I think we refer to them as `course administrators` elsewhere

nicolaasmatthijs commented on node_modules/col-assets/tests/test-assets.js:
```diff
@@ -881,9 +882,9 @@ describe('Assets', function() {
 
     /**
-     * Test that verifies that an asset can be deleted
+     * Test that verifies that an asset can be deleted by an instructor
```
I would go with the `administrator` language here as well

nicolaasmatthijs commented on node_modules/col-assets/tests/test-assets.js:
```diff
@@ -920,6 +921,58 @@ describe('Assets', function() {
+                                WhiteboardsTestUtil.assertCreateWhiteboard(client1, course, 'UC Riverside Whiteboard', [user1Me.id], function(whiteboard) {
+                                  WhiteboardsTestUtil.addAssetToWhiteboard(client1, course, asset4, whiteboard, function() {
+                                    // Verify that the creator cannot delete the used asset
+                                    AssetsTestUtil.assertDeleteAssetFails(client1, course, asset4.id, 401, function() {
```
Can we add a test case for an asset that's part of an exported whiteboard but no longer in use in an active whiteboard?

nicolaasmatthijs commented on node_modules/col-whiteboards/tests/util.js:
```diff
@@ -209,6 +209,36 @@ var addElementsToWhiteboard = module.exports.addElementsToWhiteboard = function(
+var addAssetToWhiteboard = module.exports.addAssetToWhiteboard = function(client, course, asset, whiteboard, callback) {
+  var assetElement = mockWhiteboardElement('path1.json', whiteboard.id);
+  assetElement.asset_id = asset.id;
+  DB.WhiteboardElement.upsert(assetElement).complete(function(err) {
```
Can you also add a TODO here to replace this with the websocket approach ?

nicolaasmatthijs commented on node_modules/col-whiteboards/tests/util.js:
```diff
@@ -209,6 +209,36 @@ var addElementsToWhiteboard = module.exports.addElementsToWhiteboard = function(
+ * Add an asset to a whiteboard
+ *
+ * @param  {RestClient}         client                        The REST client to make the request with
+ * @param  {Course}             course                        The Canvas course in which the user is interacting with the API
```
`asset` is not documented

nicolaasmatthijs commented on node_modules/col-whiteboards/tests/util.js:
```diff
@@ -209,6 +209,36 @@ var addElementsToWhiteboard = module.exports.addElementsToWhiteboard = function(
+ *
+ * @param  {RestClient}         client                        The REST client to make the request with
+ * @param  {Course}             course                        The Canvas course in which the user is interacting with the API
+ * @param  {Whiteboard}         whiteboard                    The whiteboard to add the elements to
```
We're only adding a single element

nicolaasmatthijs commented on node_modules/col-core/lib/db.js:
```diff
@@ -638,6 +639,10 @@ var setUpModel = function(sequelize) {
 
+  // An asset can be used as an element in multiple whiteboards. Note that this relation differs from the
+  // AssetWhiteboardElement relation, which describes elements in whiteboards that have been exported as assets
+  Asset.hasMany(WhiteboardElement, {'as': 'whiteboard_usages'});
```
Do we need a migration script for the production database?

nicolaasmatthijs replied:
That is probably correct. Defining these relationships always confuses me, so I just check the database to see if anything changed.

nicolaasmatthijs commented on node_modules/col-assets/tests/test-assets.js:
```diff
@@ -920,6 +921,58 @@ describe('Assets', function() {
     });
 
     /**
+     * Test that verifies that an asset can be deleted by a user under certain conditions
```
Can we replace this with `when the asset hasn't had any interactions yet` ?

nicolaasmatthijs commented:
Finished review. Back to @paulkerschen for follow-up

paulkerschen commented:
Follow-up pushed addressing PR comments and covering everything but the migration script. I'll need to check with @nicolaasmatthijs on convention for that, as I don't see previous migrations included in the project.

nicolaasmatthijs commented on node_modules/col-assets/lib/api.js:
```diff
@@ -907,6 +936,20 @@ var deleteAssets = module.exports.deleteAssets = function(ctx, assetIds, callbac
   });
 };
 
+var canDeleteAsset = function(user, asset) {
```
Needs JSDoc above the function

nicolaasmatthijs commented on node_modules/col-assets/lib/api.js:
```diff
@@ -64,6 +67,13 @@ var getAssetProfile = module.exports.getAssetProfile = function(ctx, id, increme
 
     asset = asset.toJSON();
     asset.whiteboard_elements = _.pluck(asset.whiteboard_elements, ['element']);
+    asset.exported_whiteboards = _.pluck(asset.exported_whiteboards, ['id', 'title']);
```
Do we need to do the `pluck` here as we're already doing `'attributes': ['title', 'id'],` below ?

paulkerschen replied:
In fact we do need the `pluck`; because the association runs through whiteboard elements, the Sequelize call returns an array of whiteboard elements for each whiteboard, and specifying `attributes` below isn't sufficient to screen them out. We could remove the `attributes` option, but then the database will be returning many more unneeded asset-specific properties.

nicolaasmatthijs commented on node_modules/col-assets/lib/api.js:
```diff
@@ -64,6 +67,13 @@ var getAssetProfile = module.exports.getAssetProfile = function(ctx, id, increme
+    asset.exported_whiteboards = _.pluck(asset.exported_whiteboards, ['id', 'title']);
+
+    // Usage in active whiteboards will prevent an asset from being deleted, but should not be kept in the asset profile
+    // once the canDelete property is set
```
It's not entirely clear to me what this comment is trying to say

nicolaasmatthijs commented on node_modules/col-assets/lib/api.js:
```diff
@@ -824,18 +851,20 @@ var deleteAsset = module.exports.deleteAsset = function(ctx, id, callback) {
     }
 
+    if (!canDeleteAsset(ctx.user, asset)) {
+      log.error({'id': id}, 'Unauthorized to delete an asset');
```
Rather than just logging an `Unauthorized` message here, can you clarify that the asset can't be deleted because it has activity associated to it?

nicolaasmatthijs commented on node_modules/col-assets/lib/api.js:
```diff
@@ -907,6 +936,20 @@ var deleteAssets = module.exports.deleteAssets = function(ctx, assetIds, callbac
+  if (user.is_admin) {
+    return true;
+  } else {
+    var isAssociatedUser = !!(_.find(asset.users, {'id': user.id}));
```
Can we get rid of the double `!` ?

nicolaasmatthijs commented on public/app/assetlibrary/item/item.html:
```diff
@@ -4,17 +4,17 @@
   <!-- TITLE -->
   <div class="row assetlibrary-item-top">
-    <div class="col-xs-{{ me.is_admin ? 5 : (canManageAsset() ? 8 : 12) }} col-sm-{{ me.is_admin ? 7 : (canManageAsset() ? 10 : 12) }}">
+    <div class="col-xs-{{ asset.canDelete ? 5 : (canManageAsset() ? 8 : 12) }} col-sm-{{ asset.canDelete ? 7 : (canManageAsset() ? 10 : 12) }}">
```
I think we've mostly been using the `can_delete` format in the REST API responses

nicolaasmatthijs replied:
cfr. `is_admin` above

nicolaasmatthijs commented on node_modules/col-whiteboards/tests/util.js:
```diff
@@ -209,6 +209,57 @@ var addElementsToWhiteboard = module.exports.addElementsToWhiteboard = function(
+    }
+  };
+  DB.WhiteboardElement.destroy(whiteboardElementOptions).complete(function(err) {
+    assert.ok(!err);
```
Can we also verify that the element doesn't exist anymore?

nicolaasmatthijs commented:
With regards to the database migration, we need to develop better migration script practices. However, for this PR, I would go with a folder inside the `scripts` folder using the format `yyyymmdd-colxxx`, containing:

- .sql file that contains required SQL query to alter database structure
- Node script that actively migrates existing data to the new schema

nicolaasmatthijs commented:
Finished review. Back to @paulkerschen for follow-up

paulkerschen commented:
PR review addressed and migration script added; back to @nicolaasmatthijs.

nicolaasmatthijs commented on scripts/20160525-col378/populate_element_asset_id.js:
```diff
@@ -0,0 +1,112 @@
+var _ = require('lodash');
+var async = require('async');
+var config = require('config');
+var CollabosphereConstants = require('col-core/lib/constants');
```
This should be split into 2 blocks. I will make this change

nicolaasmatthijs commented on scripts/20160525-col378/populate_element_asset_id.js:
```diff
@@ -0,0 +1,112 @@
+ */
+var populateElementAssetId = function(assetWhiteboardElement, callback) {
+  var assetId = assetWhiteboardElement.element.assetId;
+    if (assetId) {
```
Indentation is off. I'll fix.

nicolaasmatthijs commented on scripts/20160525-col378/populate_element_asset_id.js:
```diff
@@ -0,0 +1,112 @@
+ * @param  {AssetWhiteboardElement}       assetWhiteboardElement      The element to parse an asset id from
+ * @param  {Function}                     callback                    Standard callback function
+ * @param  {Object}                       callback.err                An error that occurred, if any
+ * @param  {Number}                       callback.assetId            Parsed asset id, if any 
```
Trailing whitespace. I'll fix.

nicolaasmatthijs commented:
Followed up and merged
